packages/dashboard/src/contexts/DashContext/state.ts

```diff
@@ -0,0 +1,122 @@
+import type { providers } from "ethers";
+import type {
+  DashboardMessageBusClient,
+  ReceivedMessageLifecycle
+} from "@truffle/dashboard-message-bus-client";
+import {
+  isDashboardProviderMessage,
+  isInvalidateMessage,
+  isLogMessage,
+  isDebugMessage
+} from "@truffle/dashboard-message-bus-common";
+import type {
+  Message,
+  DashboardProviderMessage
+} from "@truffle/dashboard-message-bus-common";
+import {
+  messageNeedsInteraction,
+  messageIsUnsupported,
+  rejectMessage,
+  confirmMessage
+} from "src/utils/dash";
+import type noticeContentType from "src/components/composed/Notice/noticeContentType";
+
+type actionType =
+  | { type: "set-client"; data: DashboardMessageBusClient }
+  | { type: "set-chain-info"; data: stateType["chainInfo"] }
+  | { type: "set-notice"; data: Partial<stateType["notice"]> }
+  | {
+      type: "handle-message";
+      data: {
+        lifecycle: ReceivedMessageLifecycle<Message>;
+        provider: providers.JsonRpcProvider;
+      };
+    };
+
+type stateType = {
+  host: string;
+  port: number;
+  client: DashboardMessageBusClient | null;
+  providerMessages: Map<
+    number,
+    ReceivedMessageLifecycle<DashboardProviderMessage>
+  >;
+  chainInfo: {
+    id: number | null;
+    name: string | null;
+  };
+  notice: {
+    show: boolean;
+    type: noticeContentType | null;
+  };
+};
+
+const initialState: stateType = {
+  host: window.location.hostname,
+  port:
+    process.env.NODE_ENV === "development"
+      ? 24012
+      : Number(window.location.port),
+  client: null,
+  providerMessages: new Map(),
+  chainInfo: {
+    id: null,
+    name: null
+  },
+  notice: {
+    show: false,
+    type: "LOADING"
+  }
+};
+
+const reducer = (state: stateType, action: actionType): stateType => {
```

Comment:
wouldn't it be better to use immer/zustand instead of this huge set of case statements/reducer code?

Reply:
It's a switch statement with 4 cases. Reducer suffices.

Reply:
your not counting the cyclomatic complexity of your last case and the several if statements embedded within that are are really a full other set of logic.

Reply:
Could you explain how a different state management lib would make it easier / cleaner to determine dashboard message type? 
```ts
if (isDashboardProviderMessage(message)) {
  // ...
} else if (isInvalidateMessage(message)) {
  // ...
} else if (isLogMessage(message)) {
  // ...
} else if (isDebugMessage(message)) {
  // ...
}
```

Reply:
Using a state management library definitely seems like overkill here... perhaps in the future, we'll change our mind, but let's not add dependencies in this PR.

Also, if we do want to move in the direction of bringing in a first-class state management solution, then we should probably stick with redux because we use it elsewhere already.

Note that if we do use a state management solution, we will likely have our work cut out for us with these action types... i.e., Redux pretty strongly demands that all actions are serializable plain objects, and `DashboardMessageBusClient` is extremely not serializable.